File: src/main/golang/leetcode/editor/cn/[211]Design Add and Search Words Data Structure.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
type WordDictionary struct {
	children [26]*WordDictionary
	isEnd    bool
}
type Tuple struct {
	node  *WordDictionary
	index int
}

func Constructor() WordDictionary {
	return WordDictionary{}
}

func (this *WordDictionary) AddWord(word string) {
	node := this
	for _, c := range word {
		idx := c - 'a'
		if node.children[idx] == nil {
			node.children[idx] = &WordDictionary{}
		}
		node = node.children[idx]
	}
	node.isEnd = true
}
func (this *WordDictionary) SearchAll(word string) (res []*WordDictionary) {
	var queue = []Tuple{{node: this, index: 0}}
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr.node == nil {
			continue
		}
		if curr.index == len(word) {
			res = append(res, curr.node)
			continue
		}
		c := word[curr.index]
		if c == '.' {
			for i := 0; i < 26; i++ {
				if curr.node.children[i] != nil {
					queue = append(queue, Tuple{node: curr.node.children[i], index: curr.index + 1})
				}
			}
		} else {
			idx := c - 'a'
			if curr.node.children[idx] != nil {
				queue = append(queue, Tuple{node: curr.node.children[idx], index: curr.index + 1})
			}
		}
	}
	return
}

func (this *WordDictionary) Search(word string) bool {

	var dfs func(i int, node *WordDictionary) bool
	dfs = func(i int, node *WordDictionary) bool {
		if node == nil {
			return false
		}
		if i == len(word) {
			return node.isEnd
		}
		curr := word[i]
		if curr == '.' {
			for j := 0; j < 26; j++ {
				if dfs(i+1, node.children[j]) {
					return true
				}
			}
		} else {
			idx := curr - 'a'
			return dfs(i+1, node.children[idx])
		}
		return false
	}
	return dfs(0, this)
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */
//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[211]Design Add and Search Words Data Structure.py
# leetcode submit region begin(Prohibit modification and deletion)
class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.trie = {}

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        node = self.trie
        for c in word:
            node = node.setdefault(c, {})
        node['#'] = '#'  # mark the end of word

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure.
        A word could contain the dot character '.' to represent any one letter.
        """
        return self._search(word, self.trie)

    def _search(self, word: str, node: dict) -> bool:
        if not word:
            return '#' in node
        if word[0] == '.':
            return any(self._search(word[1:], node[c]) for c in node if c != '#')
        else:
            if word[0] not in node:
                return False
            return self._search(word[1:], node[word[0]])
# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
# leetcode submit region end(Prohibit modification and deletion)
